### ISSUE TITLE
Testing against oldest python dependencies 

### ISSUE THREAD

**Issue by seberg** [rapidsai/build-planning#81]
E.g. for NumPy 2 it makes sense at least for some projects (cudf, cuml, cucim) to test both NumPy 2 and NumPy 1.x, however, RAPIDS currently doesn't do this.

The following PR would allow each projects to add explicit oldest dependencies to their testing workflows: https://github.com/rapidsai/shared-workflows/pull/228

- [x] [Shared workflow PR](https://github.com/rapidsai/shared-workflows/pull/228)
- [x] [Include dependency file generator in wheel image](https://github.com/rapidsai/ci-imgs/pull/163)

For each of the projects, we need then pin the interesting oldest dependencies:
- [x] **`rmm`**:
  - `numpy==1.23`, `numba=0.57`
  - [PR](https://github.com/rapidsai/rmm/pull/1613)
- [x] **`cuDF`**:
  - `numpy==1.23`, **`pandas==2.0`**, `numba=0.57`, (`fsspec=0.6.0`)
    `pyarrow==16.1.0`, `cupy==12.0.0`
  - [PR](https://github.com/rapidsai/cudf/pull/16570)
- [ ] **`raft`**:
  - `numpy=1.23`, `numba==0.57`, (`joblib==0.11`)
- [ ] **`cuML`**:
  - `numpy=1.23`, `numba==0.57`, **`scipy==1.8`**, `joblib==0.11`
- [ ] **`cuGraph`**:
  - `cupy==12.0.0`, `numpy=1.23`, `numba==0.57`, **`networkx==3.0`**, (`fsspec=0.6.0`)
- [ ] **`cuCIM`**:
  - `cupy==12.0.0`, `numpy=1.23`, **`scikit-image==0.19.0`**, (`lazy_loader==0.1`)
- [ ] **`cuSpatial`**:
  - `cupy==12.0.0`, `numpy=1.23`, **`geopandas=0.11.0`**
- [ ] **`cuxfilter`**:
  - `cupy==12.0.0`, `numpy=1.23`, `numba>=0.57`, **`geopandas=0.11.0`, `holoviews==1.16.0`, `bokeh==3.1`**, `panel==1.0`
- [ ] **`ucxx`**:
  - `numpy==1.23`, `numba=0.57`  (only inherited from RMM, maybe skip)
- [ ] **`KvikIO`**:
  - `cupy==12.0.0`, `numpy=1.23`, `zarr==?` (no range specified)
- [ ] **`wholegraph`**:
  - *probably skip dependencies seem only test dependencies not runtime.*
- [ ] **`cuVS`**:
  - `cupy==12.0.0`, `numpy==1.23`

Many of these depend on `rmm` so could e.g. have a `numba=0.57`, but maybe it is sufficient to concentrate on explicitly specified dependencies only. I bracketed a few that seemed not very relevant, although it may make sense to just put them anyway.

**Comment by jameslamb** [comment]
+1, I think it'll be great to have something like this. I've seen it be very effective in other projects.

And I specifically think we'll be happy to have it when `pandas` 3.0 comes out: https://github.com/pandas-dev/pandas/issues/57064

**Comment by dantegd** [comment]
@seberg was checking the work being done here, and it looks great! Something that also popped un my mind (thinking of cuML and Scikit-learn) is that it might be very very useful to also be able to do early forward version testing. In the `rmm` PR I see there is the `earliest` field in `dependencies.yaml`, I think having a `latest` could be just as useful, for example to test release candidate versions (or beta versions), or package versions before we fully support them. What do you think?

cc @bdice @vyasr @beckernick @betatim who might have thoughts about that as well.

**Comment by jakirkham** [comment]
@bdice do you have thoughts on this proposal? 🙂

**Comment by bdice** [comment]
Yes, I support this. This is what @seberg and I discussed when I drafted https://github.com/rapidsai/shared-workflows/pull/228.

**Comment by jakirkham** [comment]
The first 2 steps are done above. Have checked those boxes

Next step is rolling this out to projects (as described by the remaining checklist items)

Though please let me know if I've missed anything

**Comment by vyasr** [comment]
Is this something @seberg can continue working on for the rest of RAPIDS?

**Comment by seberg** [comment]
Sorry, was out for a while, so didn't pick it up again after the cudf start.  Yes, I think I can continue working on this again very soon.

**Comment by jakirkham** [comment]
Having learned from the cuDF experience, it is worth noting that we will need a collaborator from each of the libraries to help push these PRs to completion

Setting up the matrix is doable. Debugging the various RAPIDS project specific errors that come up is something best left in the hands of a project-focused developer

**Comment by vyasr** [comment]
I agree, that sounds like the right plan. I think having a PR coming from someone build-focused can help motivate the library devs to come up with the proper dependency constraints and fix any unexpected issues, so it's good to start here then hand off.

**Comment by vyasr** [comment]
Given the discussion above, I am going to close this issue. If projects want to start testing against oldest dependencies all the infrastructure is in place, they just have to update their dependencies.yaml to do that.
